ENCD-6097-allow-series-and-related-datasets-in-cart

forresttanaka opened the pull request:
Preview demo: ~https://encd-6097-series-cart-pv2-fytanaka.demo.encodedcc.org/~

I moved the `<li>` above the level of `Listing()` so that I could insert the series manager actuators inside the `<li>` on the cart-view page without having to add this actuator to the Listing components. This does however mean I had to change the `<li>` to `<div>` on all existing Listing components.

forresttanaka committed:
ENCD-6097-series-cart

zoldello commented on src/encoded/static/components/cart/cart.js:
```diff
@@ -619,32 +582,55 @@ CartSearchResultsControls.defaultProps = {
+ * @return {array} `series` with search-result series added
+ */
+const addToAccumulatingSeries = (series, currentResults) => {
+    if (currentResults['@graph'] && currentResults['@graph'].length > 0) {
```
Equivalent to - 
`currentResults['@graph']?.length > 0`

zoldello commented on src/encoded/static/components/cart/cart.js:
```diff
@@ -827,6 +813,33 @@ const processDatasetTargetList = (datasets) => {
+ */
+const processSeriesDatasets = (series, datasets) => {
+    if (series && series.length > 0) {
+        series.forEach((singleSeries) => {
```
This may be simpler if you use a filter.

forresttanaka replied:
This loop mutates the elements. I prefer to reserve `filter()` for idempotent operations.

zoldello commented on src/encoded/static/components/cart/cart.js:
```diff
@@ -827,6 +813,33 @@ const processDatasetTargetList = (datasets) => {
+            if (singleSeries.related_datasets && singleSeries.related_datasets.length > 0) {
+                singleSeries.related_datasets.forEach((relatedDataset) => {
+                    const matchingRelatedDataset = datasets.find((dataset) => relatedDataset['@id'] === dataset['@id']);
+                    if (matchingRelatedDataset) {
```
This and the child if can be combined.

zoldello commented on src/encoded/static/components/cart/remove_multiple.js:
```diff
@@ -29,7 +29,11 @@ const CartRemoveElementsModalComponent = ({ elements, closeClickHandler, removeM
             <ModalBody>
-                <p id="description-remove-selected">Remove the {elements.length} currently selected datasets from the cart. This action is not reversible.</p>
+                <p id="description-remove-selected">
+                    Remove the {elements.length} currently selected datasets from the cart. Series
```
What is elements.length === 0 or 1?

forresttanaka replied:
It can’t be 0 or this wouldn’t get called. It can certainly be 1.

zoldello commented on src/encoded/static/components/cart/remove_multiple.js:
```diff
@@ -116,19 +123,22 @@ const CartRemoveElementsComponent = ({
+            <div className="remove-multiple-control__note">
+                Any datasets belonging to a series remain after removing the selected items from the cart.
+            </div>
             {isModalVisible
```
Maybe use the ??-operator ..?

forresttanaka replied:
`isModalVisible` is a boolean, so I can use &&.

zoldello commented on src/encoded/static/libs/ui/modal.js:
```diff
@@ -286,53 +290,37 @@ ModalFooter.defaultProps = {
+     * @returns {DOM node} Injected DOM node reference for modal
+     */
+    const getModalEl = () => {
+        if (modalEl.current === null) {
```
null or falsey?

forresttanaka replied:
Actually, not sure. I’ll leave it for now since it’s existing code.

zoldello commented on src/encoded/static/libs/ui/modal.js:
```diff
@@ -343,24 +331,52 @@ export class ModalElement extends React.Component {
+
+        // Focus screen reader on the given element if specified.
+        if (focusId) {
+            const focusElement = document.getElementById(focusId);
```
Maybe use ref?

forresttanaka replied:
You can for an external component; a little bit tricky. I’ll leave this for now since it’s existing code.

zoldello commented on src/encoded/static/libs/ui/modal.js:
```diff
@@ -343,24 +331,52 @@ export class ModalElement extends React.Component {
+        document.addEventListener('keydown', handleTab, false);
+
+        // Collect all the focusable elements.
+        focusableElements.current = getModalEl().querySelectorAll('a[href], button, textarea, input[type="text"], input[type="radio"], input[type="checkbox"], select');
```
I do things like this too. I am curious about your opinions on doing this in react apps; after the sprint is over.

zoldello reviewed (dismissed):
Most of my comments are ignorable. However, you may want to consider the one with element.length === 1 or 0.

forresttanaka committed:
ENCD-6097-series-cart

Updates from code-review feedback.

forresttanaka committed:
ENCD-6097-series-cart

Gray out facets while Series tab selected. Change text of warning when removing series.

zoldello reviewed (dismissed):
Looks fine. The failing tests is a CI issue.

zoldello commented:
@hitz With visual inspection, the rebase brought up nothing concerning.
